### ISSUE TITLE
[DISCUSSION] `useSyncExternalStore` for browser API hooks

### ISSUE THREAD

**Issue by moritzsalla** [buildinamsterdam/use-match-media#12]
Not much of a feature request, just want to kick of a discussion 🙃

Since the introduction of `useSyncExternalStore`, there's a dedicated way to subscribe to browser APIs in react over useState/useEffect.

Has any of you used `useSyncExternalStore` before? Would there be worthwhile improvements to using `useSyncExternalStore` over `useState/useEffect`?

As far as I can see from a quick skim through the docs, potential improvements are:
- Improving concurrent rendering by preventing tearing, which might become an issue with our current approach once concurrent mode ships (see https://julesblom.com/writing/usesyncexternalstore#why-not-useeffect-with-usestate)
- Clear seperation of client and server logic / easier to prevent hydration issues.

Looking forward to hearing your thoughts!

**Comment by PauloMFJ** [comment]
Really cool. Had a read through that article and had no idea it was a thing! Would love to play around with it. My main question right now:
1. We're only creating a single event listener for each query, no matter how many instances are used. In this case, I wonder how `useSyncExternalStore` handles this. For example, does it create a new listener for each instance of the hook, or does it handle the event listeners under the hood as a single listener / in another way that's actually more performant?

If the latter, it would massively simplify the implementation. If the former, wonder if we could adapt it and combine our current approach with this new hook. In any case though, for the examples they provided about UI inconsistencies, seems like this is definitely something we should explore in this lib + a bunch of our code. Great suggestion 👌

**Comment by moritzsalla** [comment]
Looks like a promising hook indeed 🎉

Apart from memoizing the value like the guy mentioned, making it immuteable by parsing it as json also works. So hooking into a single event listener is possible. At least in the context of the example. 

```jsx
export const useWindowSize = () => {
  const windowSize = useSyncExternalStore(
    onWindowSizeChange,
    getWindowSizeSnapshot,
    getWindowSizeSnapshotServer,
  )

  return JSON.parse(windowSize)
}

const onWindowSizeChange = (onChange) => {
  window.addEventListener('resize', onChange)
  return () => window.removeEventListener('resize', onChange)
}

const getWindowSizeSnapshot = () => {
  return JSON.stringify({
    width: window.innerWidth,
    height: window.innerHeight,
  })
}

const getWindowSizeSnapshotServer = () => {
  return JSON.stringify({
    width: 0,
    height: 0,
  })
}
```

I'm also curious if it'd be possible to replicate our current setup of subscribing to every listener once. 

To my understanding the callbacks can all be placed within the component (no need for `useCallback` either). Just spitballing here but would it'd be a potential direction to handle the `addListener`/`removeListener` logic within the subscriber?

```jsx
const queries = new Map()

// existing logic

export const useSyncedHook = (query, initialValue) => {
  const subscriber = (onChange) => {
    subscribe(query)
    return () => unsubscribe(query)
  }

  const getServerSnapshot = () => initialValue

  const data = useSyncExternalStore(
    subscriber,
    getClientSnapshot,
    getServerSnapshot,
  )

  return data
}
```

**Comment by PauloMFJ** [comment]
Super interesting, love the examples. Definitely seems possible, granted right now we'd not really see any benefits. A change like this gets us ready for when we do. Would be keen to explore updating this package to fit this new approach. It would also give us a really good reference point for future libs.

The only downside is we'd no longer support React 17. But not sure that's really an issue at this point. Older projects could always just install an older version of the lib. Keen to see if anyone else has any other thoughts?

**Comment by MaartenBruggink** [comment]
Yeah this makes perfect sense to switch to imo, like Paulo mentioned for projects that aren't on React 18 they can use the current version. Based on the info in the docs this looks like the way forward for these kind of integrations
